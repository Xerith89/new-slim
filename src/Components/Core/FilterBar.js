import React, { Component } from 'react'
import PropTypes from 'prop-types'
import {filterByString} from '../Util/Filter'
import Pagination from './Pagination';

export default class FilterBar extends Component {

    constructor(props){
        super(props)

        this.state = this.getInitialState();
    }

    getInitialState = () => {
        const initialState = {
            filteredData:  this.props.data,
        }
        this.props.filterOptions.forEach((e) => {
            initialState[e.header] = ''
        })
        
        return initialState;
    }

    clearfilter = () => {
        this.setState(this.getInitialState());
    }

    handleChange = (event) => {
        let refiltered = this.props.data;
        //We're taking off a filter so recursively check all the others that are set
        if (event.target.value === '') {
            this.props.filterOptions.forEach((dropdown) => {
                if (dropdown.header !== event.target.name) {
                    refiltered = filterByString(refiltered, dropdown.header, this.state[dropdown.header]);
                }                
            });
           this.setState({
               filteredData: refiltered,
               [event.target.name]: event.target.value
           });
           //Setting a filter
        } else {
            this.props.filterOptions.forEach((dropdown) => {
                if (this.state[dropdown.header] !== '' && event.target.name !== dropdown.header) {
                    refiltered = filterByString(refiltered,dropdown.header,this.state[dropdown.header]);
                } else if(event.target.name === dropdown.header) {
                    refiltered = filterByString(refiltered,dropdown.header,event.target.value);
                }
            });
            this.setState({
                filteredData: refiltered,
                [event.target.name] : event.target.value
            });
        }
        
    }
   
    render() {
        return (
            <div>
                 <div id="overview" className="card m-4" style={{width: 'auto'}} data-testid="filterBar">
                 <div className="form-group m-2">
                 {this.props.filterOptions && this.props.filterOptions.map((element) => { 
                     return (<React.Fragment key={element.header}>
                                    <label className="mr-1 ml-2 mt-3 mb-2">{element.header}:</label>
                                    <select id={element.header} name={element.header} value={this.state[element.header]} onChange={this.handleChange} className="form-control-sm col-sm-2 mr-2">
                                    {element.body.map((option, j) => { 
                                    return( <option value={option} key={j}>{option}</option>)
                                    })}
                                    </select>
                            </React.Fragment>)
                            })}
                         <button onClick={this.clearfilter} className="btn btn-primary ml-3">Clear Filter</button>
                       </div>
                 </div>
                 <Pagination fetching={this.props.fetching} recordsPerPage={6} filteredData={this.state.filteredData} tableHeaderOptions={this.props.tableHeaderOptions}/>
            </div>
        )
    }
}

FilterBar.propTypes = {
    options: PropTypes.arrayOf(Object),
    fetchingTasks: PropTypes.bool,
    taskList: PropTypes.array
}
