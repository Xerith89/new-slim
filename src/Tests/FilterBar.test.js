import React from 'react';
import {render, cleanup} from '@testing-library/react';
import FilterBar from './../Components/Core/FilterBar';
import {shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({adapter: new Adapter()});

afterEach(cleanup);

const  taskList = [
    {
        id: 1,
        taskName: 'Review New Claim',
        claimSpecNo: '5500001',
        type: 'Claim', 
        assigned: 'Alex Coupe',
        priority: 'Normal',
        dueDate: '01-09-2019'
    },
    {
        id: 2,
        name: 'Review New Invoice',
        claimSpecNo: '5500000',
        type: 'Claim', 
        assigned: 'Alex Coupe',
        priority: 'Normal',
        dueDate: '01-09-2019'
    },
    {
        id: 3,
        name: 'Diary Review',
        claimSpecNo: '5500244',
        type: 'Claim', 
        assigned: 'Donald Duck',
        priority: 'Urgent',
        dueDate: '29-09-2019'
    },
    {
        id: 4,
        name: 'Loss Adjustor Report Received',
        claimSpecNo: '5500044',
        type: 'Claim', 
        assigned: 'Peter Parker',
        priority: 'Urgent',
        dueDate: '05-09-2019'
    },
]

test('TableHeader Renders To The Dom Without Spinner', () => {
    const {getByTestId} = render(<FilterBar taskList={taskList} fetchingTasks={false}/>);
    expect(getByTestId('filterBar')).toMatchSnapshot();
});

test('Sorting Function Fires On Task Name Click', () => {
    const spy = jest.fn();
    let wrapper = shallow(<TableHeader taskList={taskList} fetchingTasks={false}/>);
    wrapper.instance().handleClick = spy;
    wrapper.instance().doSort = spy;
    wrapper.instance().forceUpdate();
    wrapper.update();
    expect(wrapper.instance().handleClick).not.toHaveBeenCalled();
    wrapper.find('#id-0').simulate('click');
    expect(wrapper.instance().handleClick).toHaveBeenCalled();
    expect(wrapper.instance().handleClick).toHaveBeenCalledTimes(1);
    expect(wrapper.instance().doSort).toHaveBeenCalled();
    wrapper.unmount();
});

test('Sorting Function Fires On Spec/Claim No Click', () => {
    const spy = jest.fn();
    let wrapper = shallow(<TableHeader taskList={taskList} fetchingTasks={false}/>);
    wrapper.instance().handleClick = spy;
    wrapper.instance().doSort = spy;
    wrapper.instance().forceUpdate();
    wrapper.update();
    expect(wrapper.instance().handleClick).not.toHaveBeenCalled();
    wrapper.find('#id-1').simulate('click');
    expect(wrapper.instance().handleClick).toHaveBeenCalled();
    expect(wrapper.instance().handleClick).toHaveBeenCalledTimes(1);
    expect(wrapper.instance().doSort).toHaveBeenCalled();
    wrapper.unmount();
});

test('Sorting Function Fires On Type Click', () => {
    const spy = jest.fn();
    let wrapper = shallow(<TableHeader taskList={taskList} fetchingTasks={false}/>);
    wrapper.instance().handleClick = spy;
    wrapper.instance().doSort = spy;
    wrapper.instance().forceUpdate();
    wrapper.update();
    expect(wrapper.instance().handleClick).not.toHaveBeenCalled();
    wrapper.find('#id-2').simulate('click');
    expect(wrapper.instance().handleClick).toHaveBeenCalled();
    expect(wrapper.instance().handleClick).toHaveBeenCalledTimes(1);
    expect(wrapper.instance().doSort).toHaveBeenCalled();
    wrapper.unmount();
});

test('Sorting Function Fires On Assigned To Click', () => {
    const spy = jest.fn();
    let wrapper = shallow(<TableHeader taskList={taskList} fetchingTasks={false}/>);
    wrapper.instance().handleClick = spy;
    wrapper.instance().doSort = spy;
    wrapper.instance().forceUpdate();
    wrapper.update();
    expect(wrapper.instance().handleClick).not.toHaveBeenCalled();
    wrapper.find('#id-3').simulate('click');
    expect(wrapper.instance().handleClick).toHaveBeenCalled();
    expect(wrapper.instance().handleClick).toHaveBeenCalledTimes(1);
    expect(wrapper.instance().doSort).toHaveBeenCalled();
    wrapper.unmount();
});

test('Sorting Function Fires On Priority Click', () => {
    const spy = jest.fn();
    let wrapper = shallow(<TableHeader taskList={taskList} fetchingTasks={false}/>);
    wrapper.instance().handleClick = spy;
    wrapper.instance().doSort = spy;
    wrapper.instance().forceUpdate();
    wrapper.update();
    expect(wrapper.instance().handleClick).not.toHaveBeenCalled();
    wrapper.find('#id-4').simulate('click');
    expect(wrapper.instance().handleClick).toHaveBeenCalled();
    expect(wrapper.instance().handleClick).toHaveBeenCalledTimes(1);
    expect(wrapper.instance().doSort).toHaveBeenCalled();
    wrapper.unmount();
});

test('Sorting Function Fires On Due Date Click', () => {
    const spy = jest.fn();
    let wrapper = shallow(<TableHeader taskList={taskList} fetchingTasks={false}/>);
    wrapper.instance().handleClick = spy;
    wrapper.instance().doSort = spy;
    wrapper.instance().forceUpdate();
    wrapper.update();
    expect(wrapper.instance().handleClick).not.toHaveBeenCalled();
    wrapper.find('#id-4').simulate('click');
    expect(wrapper.instance().handleClick).toHaveBeenCalled();
    expect(wrapper.instance().handleClick).toHaveBeenCalledTimes(1);
    expect(wrapper.instance().doSort).toHaveBeenCalled();
    wrapper.unmount();
});