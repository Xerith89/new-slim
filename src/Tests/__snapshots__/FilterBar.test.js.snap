// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterBar Renders To The Dom 1`] = `
<div
  class="card m-4"
  data-testid="filterBar"
  id="overview"
  style="width: auto;"
>
  <div
    class="form-group m-2"
  >
    <label
      class="mr-1 ml-2 mt-3 mb-2"
    >
      Type
      :
    </label>
    <select
      class="form-control-sm col-sm-2 mr-2"
      id="Type"
      name="Type"
    >
      <option
        value=""
      />
      <option
        value="Claim"
      >
        Claim
      </option>
      <option
        value="Specification"
      >
        Specification
      </option>
    </select>
    <label
      class="mr-1 ml-2 mt-3 mb-2"
    >
      Assigned
      :
    </label>
    <select
      class="form-control-sm col-sm-2 mr-2"
      id="Assigned"
      name="Assigned"
    >
      <option
        value=""
      />
      <option
        value="Alex Coupe"
      >
        Alex Coupe
      </option>
      <option
        value="Donald Duck"
      >
        Donald Duck
      </option>
      <option
        value="Peter Parker"
      >
        Peter Parker
      </option>
    </select>
    <label
      class="mr-1 ml-2 mt-3 mb-2"
    >
      Priority
      :
    </label>
    <select
      class="form-control-sm col-sm-2 mr-2"
      id="Priority"
      name="Priority"
    >
      <option
        value=""
      />
      <option
        value="Normal"
      >
        Normal
      </option>
      <option
        value="Urgent"
      >
        Urgent
      </option>
    </select>
    <button
      class="btn btn-primary ml-3"
    >
      Clear Filter
    </button>
  </div>
</div>
`;
